### PR TITLE
Align AvgPool ceil_mode on last value to torch

### LINKED ISSUE #16203: `count_include_pad` calculates the the most right output with dividing it by kernel size regardless it's not padded.
### Describe the issue

From https://github.com/onnx/onnx/issues/5276
`count_include_pad` calculates the most right output with dividing it by kernel size regardless it's not padded. I think it might be a bug that the original intention should be divided by the number of remaining pixels (including padding).

### To reproduce

I am sorry, but it's easier to repro with onnx-script eager mode (backend ORT)

```python
import numpy as np
import onnxscript
from onnxscript.onnx_opset import opset18 as op

x = np.array([[[ 2.0903,  4.6493,  1.6320, -3.2051,  4.6975,  4.7296,  3.3653,-1.5815, -2.3832],
               [ 0.9628, -1.5899, -2.6820,  5.7529,  7.7346, -0.8910, -2.0151,0.1313, -0.5374]]]).astype(np.float32)


@onnxscript.script(default_opset=op)
def avg_pool(x):
    result = op.AveragePool(x, kernel_shape=[7], strides=[3], pads=[3,3], ceil_mode=True, count_include_pad=True)
    return result

print(avg_pool(x))
# [[[ 0.73807144  2.5655572   0.8032287  -0.08562858]
#   [ 0.34911433  1.0389      1.4536142  -0.34588572]]]
```

In PyTorch, it is divided by the number of remaining pixels (including padding):

```python
import torch

a = torch.tensor([[[ 2.0903,  4.6493,  1.6320, -3.2051,  4.6975,  4.7296,  3.3653,-1.5815, -2.3832],[ 0.9628, -1.5899, -2.6820,  5.7529,  7.7346, -0.8910, -2.0151,0.1313, -0.5374]]])
p = torch.nn.AvgPool1d((7,), (3,), (3,), ceil_mode=True, count_include_pad=True)
p(a)

# tensor([[[ 0.7381,  2.5656,  0.8032, -0.0999],
#               [ 0.3491,  1.0389,  1.4536, -0.4035]]])

``` 

To summary, we can take a specific looke on the output of the first row. The reason that they are mis matched is because they are divided by different number of remaing pixels:

In the last window slide,
PyTorch: sum([3.3653,-1.5815, -2.3832, 0, 0, 0]) / 6 = -0.0999
ORT: sum([3.3653,-1.5815, -2.3832, 0, 0, 0]) / 7 = -0.08562858

### Urgency

_No response_

### Platform

Linux

### OS Version

20.04.5

### ONNX Runtime Installation

Built from Source

### ONNX Runtime Version or Commit ID

NIghtly

### ONNX Runtime API

Python

### Architecture

X64

### Execution Provider

Default CPU

### Execution Provider Library Version

_No response_

### PR DESCRIPTION
**Pull request by titaiwangms**
Fix #16203

Previous to this PR, if `ceil_mode` is on, the calculation of a value would divide the kernel size, even if remaining pixels is less than the kernel size, which causes the difference in this operator between ORT and torch.

However, this fix only applies to the change in #15597, which only supports AvgPool since 19. The older opset version is remain the same, as it's using mlas files.

Also, the PR fixes the shape mismatch caused by sliding window starting from padding. More detail: https://github.com/onnx/onnx/pull/6650 (And this PR is also validated with the tests added in https://github.com/onnx/onnx/pull/6650)

### CODE REVIEW

File: onnxruntime/test/providers/cpu/nn/pool_op_test.cc [hunk 1]
```diff
@@ -967,6 +967,31 @@ TEST(PoolTest, AveragePool_19_dilation_2d) {
   test.Run(OpTester::ExpectResult::kExpectSuccess, "", {kTensorrtExecutionProvider, kAclExecutionProvider, kOpenVINOExecutionProvider});
 }
 
+TEST(PoolTest, AveragePool_19_ceil_count_include_pad_1d) {
+  // TODO: Unskip when fixed #41968513
```
Comment (justinchuby) [inline onnxruntime/test/providers/cpu/nn/pool_op_test.cc:971]: is #41968513 a valid issue?

File: onnxruntime/core/providers/cpu/nn/pool_functors.h [hunk 2]
```diff
@@ -406,6 +406,7 @@ struct AveragePool1DTask final {
     for (int64_t ph = 0; ph < pooled_height; ++ph) {
       int64_t hstart = ph * stride_h - pads[0];
       int64_t hend = hstart + kernel_shape[0] * dilation_h;
+      hend = std::min(hend, height + pads[1]);
```
Comment (justinchuby) [inline onnxruntime/core/providers/cpu/nn/pool_functors.h:409]: Is the same needed for other providers as well?

**Review by justinchuby — APPROVED** [review]

File: onnxruntime/test/providers/cpu/nn/pool_op_test.cc [hunk 3]
```diff
@@ -967,6 +967,31 @@ TEST(PoolTest, AveragePool_19_dilation_2d) {
...
+  OpTester test("AveragePool", 19);
+
+  test.AddAttribute("auto_pad", "");
+  test.AddAttribute("strides", std::vector<int64_t>{3});
+  test.AddAttribute("pads", vector<int64_t>{3, 3});
+  test.AddAttribute("kernel_shape", vector<int64_t>{7});
+  test.AddAttribute("ceil_mode", (int64_t)1);
+  test.AddAttribute("count_include_pad", (int64_t)1);
+
+  std::vector<float> x_vals = {2.0903, 4.6493, 1.6320, -3.2051, 4.6975, 4.7296, 3.3653, -1.5815, -2.3832, 0.9628, -1.5899, -2.6820, 5.7529, 7.7346, -0.8910, -2.0151, 0.1313, -0.5374};
```
Comment (liqunfu) [inline onnxruntime/test/providers/cpu/nn/pool_op_test.cc:985]: 2.0903f

File: onnxruntime/test/providers/cpu/nn/pool_op_test.cc (hunk 1 above)
Comment (titaiwangms) [reply onnxruntime/test/providers/cpu/nn/pool_op_test.cc:971]: I think so. I checked there are other 14 test cases having this comment. I guess they have it somewhere.

File: onnxruntime/core/providers/cpu/nn/pool_functors.h (hunk 2 above)
Comment (titaiwangms) [reply onnxruntime/core/providers/cpu/nn/pool_functors.h:409]: I only have cpu on my machine, follow #15597, I would only update this on CPU. I am thinking riase another issue to ask for help from core runtime team to apply this to all versions and providers.

File: onnxruntime/core/providers/cpu/nn/pool_functors.h [hunk 4]
```diff
@@ -461,6 +462,7 @@ struct AveragePool2DTask final {
...
```
Comment (xadupre) [inline onnxruntime/core/providers/cpu/nn/pool_functors.h:465]: Shoud it be added to wend as well?

Comment (titaiwangms) [reply onnxruntime/core/providers/cpu/nn/pool_functors.h:465]: Also updated to 2d and 3d

**Comment by justinchuby** [comment]
Was this discrepancy clarified by the ONNX spec? Was it due to an implementation oversight of the spec? Hopefully the spec itself was not different than torch?

**Comment by gramalingam** [comment]
I was trying a few things out: here's some strange behavior: `AveragePool(x, kernel_shape=[2], strides=[2], pads=[1, 1])` seems to produce a strange behavior ... the left-padding behavior is different from the right-padding behavior. Is it the same in torch?

**Review by justinchuby — DISMISSED** [review]

**Comment by titaiwangms** [comment]
Hi @baijumeswani Yes. But this PR only works on Pooling op after opset18. Could you help us to add this supports on older versions as well?

**Comment by titaiwangms** [comment]
@liqunfu @gramalingam @xadupre @justinchuby PTAL

**Review by justinchuby — APPROVED** [review]
